cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -137,6 +148,157 @@ func createKubeDNSAddon(deploymentBytes, serviceBytes []byte, client clientset.I
+	if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), clusterroleBytes, coreDNSClusterRoles); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS clusterroles %v", err)
+	}
+
+	// Create the Clusterroles for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateClusterRole(client, coreDNSClusterRoles); err != nil {
+		return err
+	}
+
+	coreDNSClusterRolesBinding := &rbac.ClusterRoleBinding{}
+	if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), clusterrolebindingBytes, coreDNSClusterRolesBinding); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS clusterrolebindings %v", err)
+	}
+
+	// Create the Clusterrolebindings for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateClusterRoleBinding(client, coreDNSClusterRolesBinding); err != nil {
+		return err
+	}
+
+	coreDNSServiceAccount := &v1.ServiceAccount{}
+	if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), serviceAccountBytes, coreDNSServiceAccount); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS configmap %v", err)
+	}
+
+	// Create the ConfigMap for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateServiceAccount(client, coreDNSServiceAccount); err != nil {
+		return err
+	}
+
+	coreDNSDeployment := &apps.Deployment{}
+	if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), deploymentBytes, coreDNSDeployment); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS deployment %v", err)
+	}
+
+	// Create the Deployment for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateDeployment(client, coreDNSDeployment); err != nil {
+		return err
+	}
+
+	coreDNSService := &v1.Service{}
+	if err := kuberuntime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), serviceBytes, coreDNSService); err != nil {
+		return fmt.Errorf("unable to decode coreDNS service %v", err)
+	}
+
+	// Can't use a generic apiclient helper func here as we have to tolerate more than AlreadyExists.
+	if _, err := client.CoreV1().Services(metav1.NamespaceSystem).Create(coreDNSService); err != nil {
+		// Ignore if the Service is invalid with this error message:
+		// 	Service "CoreDNS" is invalid: spec.clusterIP: Invalid value: "10.96.0.10": provided IP is already allocated
+
+		if !apierrors.IsAlreadyExists(err) && !apierrors.IsInvalid(err) {
+			return fmt.Errorf("unable to create a new CoreDNS service: %v", err)
+		}
+
+		if _, err := client.CoreV1().Services(metav1.NamespaceSystem).Update(coreDNSService); err != nil {
+			return fmt.Errorf("unable to create/update the CoreDNS service: %v", err)
+		}
+	}
+	return nil
+}
+
+// convSubnet fetches the serviceCIDR and modifies the mask to the nearest class
+// CoreDNS requires CIDR notations for reverse zones as classful. Resolved in future version of CoreDNS.
+func convSubnet(cidr string) (serviceCIDR string) {
```

Comment:
What would happen in case of IPv6 here ? Are we at all supporting IPv6 in combination with CoreDNS ?

Reply:
CoreDNS supports IPv6 and would support here a reverse zone in IPv6 format. 
CoreDNS within k8s had not been tested in IPv6 mode as k8s does not support yet. 
We have to define a sequence of operations : check-in all IPv4 support first, then handle validation of  CoreDNS within k8s in IPv6 mode - and then update kubeadm for Ipv6 support of coreDNS.
@kad : can we do in 2 PRs for kubeadm or you prefer all in one ?

Reply:
I'm ok to have ipv6 support in separate PR, but then it would be good if issue about that will be filed in kubeadm repo, with appropriate milestone and in code have some comment `// TODO: issue 12345...`. Just for not to forget about it.

Reply:
FWIW, Once the install version of CoreDNS is bumped to 1.0.0 (not released yet), this method is no longer needed and should be removed.

Reply:
@miekg maybe we should release 1.0.0 soon so we can just clean this up now.

Reply:
we can do a 0.9.10 ? like today?